### DISCUSSION TITLE
validateOnly Changes + Regression

### DISCUSSION THREAD

**Discussion by stevenmond** [livewire/livewire#4278] [7. Bugs (archived)] [answered]
It looks like https://github.com/livewire/livewire/pull/4188 caused a few changes to how validation works.

First, the change to 
```
$rulesForField = collect($rules)->filter(function ($rule, $fullFieldKey) use ($field) {
             return str($field)->is($fullFieldKey) || str($fullFieldKey)->startsWith($field);
        })->toArray();
```
causes unrelated rules for fields with names of the same prefix to be returned due to the 'startsWith' function. For example, if you have rules/fields 'image' and 'image_alt' and you try to validate 'image', both rules will be returned for validation.

Second (although this may be intentional given the naming of the function), is that if you have 2 fields that are dependent upon each other (so using a ```required_with``` or ```required_without``` validation for example), due to the ```HydratePublicProperties::extractData($data, $processedRules, [])``` call, the fields in the object that's being validated against will be cut down to the single field you're trying the validate. The original version of this threw the whole object at the validator, so it had access to all fields and could handle those types of validations.

**Comment by joshhanley** [comment]
@stevenmond can you share an example set of rules, properties and validation call so I can reproduce this? Thanks!

**Comment by stevenmond** [comment]
@joshhanley Example component. Let me know if you need any other examples

```
use Livewire\Component;

class ValidatesOnlyTestComponent extends Component
{
    public string $image = '';
    public string $image_alt = '';
    public string $image_url = '';

    public $rules = [
        'image' => 'required_without:image_url|string',
        'image_alt' => 'required|string',
        'image_url' => 'required_without:image|string'
    ];

    public function mount($image, $imageAlt, $imageUrl = '')
    {
        $this->image = $image;
        $this->image_alt = $imageAlt;
        $this->image_url = $imageUrl;
    }

    public function runValidation()
    {
        $this->validate();
    }

    public function runValidateOnly($propertyName)
    {
        $this->validateOnly($propertyName);
    }

    public function runResetValidation()
    {
        $this->resetValidation();
    }

    public function render()
    {
        return view('null-view');
    }
}
```

Example test (fails on current version, works with the prior code)

```
use Illuminate\Support\Str;
use Livewire\Livewire;

class ValidatesOnlyTest extends TestCase
{
    /** @test */
    public function a_set_of_items_will_validate_individually()
    {
        if (version_compare(PHP_VERSION, '7.4', '<')) {
            $this->markTestSkipped('Typed Property Initialization not supported prior to PHP 7.4');
        }

        require_once __DIR__.'/ValidatesOnlyStub.php';

        Livewire::test(ValidatesOnlyTestComponent::class, ['image' => 'image', 'imageAlt' => 'This is an image'])
            ->call('runValidateOnly', 'image_alt')
            ->assertHasNoErrors(['image_alt', 'image_url', 'image'])
            ->call('runValidateOnly', 'image_url')
            ->assertHasNoErrors(['image', 'image_url', 'image_alt'])
            ->call('runValidateOnly', 'image')
            ->assertHasNoErrors(['image', 'image_url', 'image_alt']);
    }
}
```

Also on the first bit of code
```
  $rules = ['image' => 'required|string', 'image_alt' => 'required|string'];
  $field = 'image';
  $rulesForField = collect($rules)->filter(function ($rule, $fullFieldKey) use ($field) {
             return str($field)->is($fullFieldKey) || str($fullFieldKey)->startsWith($field);
        })->toArray();
```

outputs

```
[
  "image" => "required|string",
  "image_alt" => "required|string"
]
```

instead of
```
[
  "image" => "required|string"
]
```

**Comment by joshhanley** [comment]
@stevenmond thanks for this! 

A previous PR #3596 implemented the ability to pass a specific item in a collection (that uses a wildcard in the rule) into validateOnly (like `users.1.name`), and what they were doing was filtering down to that specific rule and changing the wildcards to specific indexs.

The problem was that custom attributes no longer worked, as the rule being passed into the validator was `users.1.name`, but custom attributes were looking for `users.*.name`.

So we had to change the strategy, now all the rules are still present, and instead we've filtered down the data that gets passed into the validator.

But what is failing now, is rules that depend on other data, so this is what is being passed into the validator

```php
$validator = Validator::make($data, $rulesForField, $messages, $attributes);
```

![image](https://user-images.githubusercontent.com/882837/143316966-e242be1a-7900-498d-8907-9569063d96ee.png)

Hence it is failing as "image_url" is empty, but it's saying it should be required if image is empty/ non existant, and as there is now no data for "image" the rule is activating.

So I guess to fix this, we have to change back to the way the rules are being filtered down to the specific rule (but with wildcards still in place), and then also filter the data but only the data for the property that matches the rule, and leave everything else intact.

**Reply by reppair** [reply]
> So I guess to fix this, we have to change back to the way the rules are being filtered down to the specific rule (but with wildcards still in place), and then also filter the data but only the data for the property that matches the rule, and leave everything else intact.

@joshhanley if it is just a matter of time and resource for this to be changed, then I can try to do it, its about time I start contributing where I can to my favorite open-source projects :)

**Reply by joshhanley** [reply]
@reppair yeah you're more than welcome to give it a go! I won't get a chance to look into it until Wednesday (Sun morning here at the moment). 🙂

All I have done is added a failing test from the code Steven shared, you can see the branch I started here https://github.com/joshhanley/livewire/tree/validate-only-fix

**Reply by reppair** [reply]
Aghh.. overwhelming amounts of work here, not sure when I can get to this.. hopefully before you :)

**Reply by joshhanley** [reply]
@reppair not sure if you got a chance to have a look at this, but I'm going to have a look at it today and hopefully get a fix done so we can release it 🙂

**Comment by mnabialek** [comment]
As mentioned here https://github.com/livewire/livewire/pull/4188#issuecomment-978993017 the problem is also when you have validation for `files` and `files.*` and for any reason you want to run validation only for `files` using this method. Previously it was possible and after this change it's not

**Comment by joshhanley** [comment]
Have submitted a fix for this in PR #4319.

@stevenmond @reppair @mnabialek can you test this if you get a chance and confirm it's all working for you?

Thanks!

**Reply by stevenmond** [reply]
Looks good on my end

**Reply by reppair** [reply]
Same here, just checked out your branch and required it in my project, it works for me as well.

Tested rules: `lt`, `lte`, `gt`, `gte` and `same`.

**Reply by joshhanley** [reply]
Great! Thanks for testing, much appreciated! 🙂

**Reply by mnabialek** [reply]
Looks good for me with 2.8.1 release

**Reply by joshhanley** [reply]
Great thanks!

**Comment by joshhanley** [comment, marked as answer]
The fix for this has been released in v2.8.1

**Reply by stevenmond** [reply]
Thanks for the fix. Everything looks good on my end.

**Reply by reppair** [reply]
Thanks!

**Reply by joshhanley** [reply]
Awesome!
